### PR TITLE
(proof-new) Distinguish pre vs post rewrites in term conversion proof generator

### PR DESCRIPTION
**Pull request by ajreynol**
This is work towards resolving two kinds of failures on proof-new:

(1) Functional issues with proofs from the rewriter, in particular when a term pre-rewrites and post-rewrites to different things,
(2) Conversion issues in theory-preprocessing, where all steps are assumed to be post-rewrites but some are in fact pre-rewrites.  This leads to the term conversion proof generator proving something different than what is expected.

A followup PR will simplify and fix proofs for theory-preprocessing.

### CODE REVIEW

File: src/theory/theory_preprocessor.cpp [hunk 1]
```diff
@@ -431,7 +432,7 @@ Node TheoryPreprocessor::rewriteWithProof(Node term)
       Trace("tpp-debug") << "TheoryPreprocessor: addRewriteStep (rewriting) "
                          << term << " -> " << termr << std::endl;
       // always use term context hash 0 (default)
-      d_tpg->addRewriteStep(term, termr, PfRule::REWRITE, {}, {term});
+      d_tpg->addRewriteStep(term, termr, PfRule::REWRITE, {}, {term}, false);
```
Comment (HanielB) [inline src/theory/theory_preprocessor.cpp:435]: ditto

File: src/theory/theory_preprocessor.cpp [hunk 2]
```diff
@@ -480,15 +482,19 @@ Node TheoryPreprocessor::preprocessWithProof(Node term)
...
                          << term << " -> " << termr << std::endl;
       // small step trust
-      d_tpg->addRewriteStep(
-          term, termr, PfRule::THEORY_PREPROCESS, {}, {term.eqNode(termr)});
+      d_tpg->addRewriteStep(term,
+                            termr,
+                            PfRule::THEORY_PREPROCESS,
+                            {},
+                            {term.eqNode(termr)},
+                            false);
```
Comment (HanielB) [inline src/theory/theory_preprocessor.cpp:497]: ditto

File: src/theory/rewriter.cpp [hunk 3]
```diff
@@ -386,7 +386,8 @@ Node Rewriter::rewriteTo(theory::TheoryId theoryId,
                                  cached,
                                  PfRule::TRUST_REWRITE,
                                  {},
-                                 {eq});
+                                 {eq},
+                                 false);
```
Comment (HanielB) [inline src/theory/rewriter.cpp:390]: that last argument is the default. Moreover, also maybe add some documentation to say why this is a post-rewrite? Does not seem so clear to me.

**Review by HanielB — APPROVED** [review]
LGTM! A few requests for doc but that's it.

I'm grad that in the end it was not a very invasive change. :)

File: src/theory/theory_preprocessor.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/theory_preprocessor.cpp:435]: This "false" is wrong (see FIXME) :)

File: src/theory/theory_preprocessor.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/theory_preprocessor.cpp:497]: Ditto

File: src/theory/rewriter.cpp (hunk 3 above)
Comment (ajreynol) [reply src/theory/rewriter.cpp:390]: Will add

**Pull request merged**